src/middleware/TestingHandler.ts

```diff
@@ -0,0 +1,265 @@
+			const response = new Response(responseBody, init);
+
+			return response;
+		} catch (error) {
+			throw error;
+		}
+	}
+
+	/**
+	 * @private
+	 * Fetches a random status code for the RANDOM mode from the predefined array
+	 * @param {string} requestMethod - the API method for the request
+	 * @returns the random status code
+	 */
+	private getRandomStatusCode(requestMethod: string): number {
+		try {
+			// returns random status code for the random method from the array present
+			const statusCodeArray: number[] = methodStatusCode[requestMethod] as number[];
+			return statusCodeArray[Math.floor(Math.random() * statusCodeArray.length)];
+		} catch (error) {
+			throw error;
+		}
+	}
+
+	/**
+	 * @private
+	 * To fetch the relative URL out of the complete URL
+	 * @param {RegExp} pattern - the regex pattern for the URL
+	 * @param {string} urlMethod - the complete URL
+	 * @returns the relative URL
+	 */
+	private getRelativeURL(pattern: RegExp, urlMethod: string): string {
+		// just helps in getting the relative URL from the complete url, just to match the url as in manual map
+		const relativeURL = urlMethod.replace(pattern, "");
+		return relativeURL;
+	}
+
+	/**
+	 * @private
+	 * To fetch the status code from the map(if needed), then returns response by calling createResponse
+	 * @param {TestingHandlerOptions} testingHandlerOptions - The TestingHandlerOptions object
+	 * @param {string} requestURL - the URL for the request
+	 * @param {string} requestMethod - the API method for the request
+	 * @returns Response object
+	 */
+	private setStatusCode(testingHandlerOptions: TestingHandlerOptions, requestURL: string, requestMethod: string) {
+		try {
+			// setting some random status code
+			testingHandlerOptions.statusMessage = "Some error happened";
+			if (testingHandlerOptions.testingStrategy === TestingStrategy.MANUAL) {
+				// inside the Manual mode
+				if (testingHandlerOptions.statusCode === undefined) {
+					// manual mode with no status code, can be a global level or request level without statusCode
+					try {
+						// chacking Manual Map for exact match
+						const urlMethod = this.getRelativeURL(new RegExp("http(s)://graph.microsoft.com/[^/]*", "g"), requestURL);
+						testingHandlerOptions.statusCode = this.manualMap.get(urlMethod).get(requestMethod);
+					} catch (error) {
+						// checking for regex match if exact match doesn't work
+						const urlMethod = this.getRelativeURL(new RegExp("http(s)://graph.microsoft.com/[^/]*", "g"), requestURL);
+						this.manualMap.forEach((value: Map<string, number>, key: string) => {
+							const regexUrl = new RegExp(key);
+							if (regexUrl.test(urlMethod)) {
+								testingHandlerOptions.statusCode = this.manualMap.get(key).get(requestMethod);
+							}
+						});
+
+						// Case of redirection or request url not in map
+						if (testingHandlerOptions.statusCode === undefined) {
+							if (requestURL === this.redirectURL) {
```

Comment:
what if the user provided requestURL is exactly same as the redirectURL that you have ? we need to generate randomized url for the redirecturl so that we can minimize the collision.